Optimize SVGSMILElement::addInstanceTime

Ahmad-S792 opened the pull request:
#### b6d8303933b5a0ff44ecf52f0ce6120429725f1c
<pre>
Optimize SVGSMILElement::addInstanceTime

<a href="https://bugs.webkit.org/show_bug.cgi?id=276862">https://bugs.webkit.org/show_bug.cgi?id=276862</a>

Reviewed by Darin Adler.

Merge: <a href="https://github.com/chromium/chromium/commit/a3c9a66d306935f8b0dd9f49d7516aaf1be5955d">https://github.com/chromium/chromium/commit/a3c9a66d306935f8b0dd9f49d7516aaf1be5955d</a>

Find the proper spot to insert the new instance time at directly and
insert it there rather than adding it to the end and std::sort&apos;ing.
Inline sortTimeList into the one remaining caller.

* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::SVGSMILElement::parseBeginOrEnd):
(WebCore::insertSorted):
(WebCore::SVGSMILElement::addInstanceTime):
(WebCore::sortTimeList): Deleted.

Canonical link: <a href="https://commits.webkit.org/281237@main">https://commits.webkit.org/281237@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e4ff19b7a578b0b04881e1758a281c901bf7c3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11673 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63087 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9581 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61267 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46141 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9812 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48077 "") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6849 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61168 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35986 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51203 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28904 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32708 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8465 "Found 1 new test failure: imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-report-only-sends-reports-on-violation.https.sub.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8585 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54664 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8745 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64782 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3068 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8684 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55399 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3081 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51206 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/55513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13121 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2570 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34297 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35380 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36466 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35125 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -751,14 +746,20 @@ SMILTime SVGSMILElement::simpleDuration() const
     auto& list = beginOrEnd == Begin ? m_beginTimes : m_endTimes;
-    list.append(SMILTimeWithOrigin(time, origin));
-    sortTimeList(list);
+    insertSorted(list, timeWithOrigin);
```
Unclear why we are adding a local variable here. Would be better like this:
```
insertSorted(list, SMILTimeWithOrigin(time, origin));
```
Or without local variables at all:
```
insertSorted(beginOrEnd == Begin ? m_beginTimes : m_endTimes, SMILTimeWithOrigin(time, origin));
```

Ahmad-S792 replied:
I opted for later to remove local variables. :-)

darinadler commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -751,14 +746,20 @@ SMILTime SVGSMILElement::simpleDuration() const
+static void insertSorted(Vector<SMILTimeWithOrigin>& list, SMILTimeWithOrigin time)
+{
+    auto position = std::lower_bound(list.begin(), list.end(), time);
+    list.insert(position - list.begin(), time);
```
Probably valuable to assert the list is already sorted in debug builds, as long as no one turns it on for release builds and ruins the optimization. I also don’t think the local variable is particularly helpful. I would do something like one of these:
```
ASSERT(std::is_sorted(list.begin(), list.end());
list.insert(std::lower_bound(list.begin(), list.end(), time) - list.begin(), time);
```
or
```
ASSERT(std::is_sorted(list.begin(), list.end());
auto position = std::lower_bound(list.begin(), list.end(), time) - list.begin();
list.insert(position, time);
```

Ahmad-S792 replied:
I opted for first option. Bit more cleaner. :-)

darinadler reviewed (approved):
Looks good. Some small style comments.

Commit:
Optimize SVGSMILElement::addInstanceTime

https://bugs.webkit.org/show_bug.cgi?id=276862

Reviewed by Darin Adler.

Merge: https://github.com/chromium/chromium/commit/a3c9a66d306935f8b0dd9f49d7516aaf1be5955d

Find the proper spot to insert the new instance time at directly and
insert it there rather than adding it to the end and std::sort'ing.
Inline sortTimeList into the one remaining caller.

* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::SVGSMILElement::parseBeginOrEnd):
(WebCore::insertSorted):
(WebCore::SVGSMILElement::addInstanceTime):
(WebCore::sortTimeList): Deleted.

Canonical link: https://commits.webkit.org/281237@main
